Prevent dust outputs from being created during the trade process

jmacxx opened the pull request:
This change fixes an issue whereby dust change outputs are inadvertently created during the trading process, unbeknownst to the user.  The dust outputs cause the Bitcoin node to reject the
transaction and the trade then becomes stuck.

The solution taken here is to detect a dust TXO during the trade process and remove it from the transaction before broadcasting.  This applies in both making an offer and/or taking an existing offer.  When a dust output is prevented, it will be noted in the log as follows:

![image](https://user-images.githubusercontent.com/47253594/77453644-82e12700-6dc5-11ea-83cf-eb2227c6275d.png)


Related to #4039

Commit:
Remove dust outputs created during the trade process

This change fixes an issue whereby dust change outputs are
inadvertently created during the trading process, unbenownst to the
user.  The dust outputs cause the Bitcoin node to reject the
transaction and the trade then becomes stuck.

The solution taken here is to detect a dust TXO during the trade
process and remove it from the transaction before broadcasting.

Related to #4039

ghost commented:
~~I neglected to mention that this initial commit only handles the path taken when BTC is used for the trading fee; when BSQ is used the transaction is built by a different routine so some work will be needed to check for dust in the same way there.~~

~~Also, a bit sloppy that I check the dust limit using a hard coded 546 sats value; `preferences.getIgnoreDustThreshold()` could be used which equates to 600 sats currently.~~

[edit] both these issues have now been addressed.

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/TradeWalletService.java:
```diff
@@ -1216,4 +1219,25 @@ private void applyLockTime(long lockTime, Transaction tx) {
+
+    private boolean removeDust(Transaction tx) {
+        List<TransactionOutput> txos = tx.getOutputs();
+        List<TransactionOutput> toKeep = new ArrayList<TransactionOutput>();
```
```suggestion
        List<TransactionOutput> toKeep = new ArrayList<>();
```

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/TradeWalletService.java:
```diff
@@ -1216,4 +1219,25 @@ private void applyLockTime(long lockTime, Transaction tx) {
     }
+
+    private boolean removeDust(Transaction tx) {
+        List<TransactionOutput> txos = tx.getOutputs();
```
More descriptive variable names are preferred...

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/TradeWalletService.java:
```diff
@@ -1216,4 +1219,25 @@ private void applyLockTime(long lockTime, Transaction tx) {
+    private boolean removeDust(Transaction tx) {
+        List<TransactionOutput> txos = tx.getOutputs();
+        List<TransactionOutput> toKeep = new ArrayList<TransactionOutput>();
+        for (TransactionOutput zz: txos) {
```
...especially names like `zz` are very hard to follow except in the most simple cases. Calling it `output` or `txOutput` to make it more readable is preferred.

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/TradeWalletService.java:
```diff
@@ -1216,4 +1219,25 @@ private void applyLockTime(long lockTime, Transaction tx) {
+                toKeep.add(zz);
+            }
+            else {
+                log.info("removing dust TXO = {}", zz.getValue().toFriendlyString());
```
A more verbose log will help log readers.

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/TradeWalletService.java:
```diff
@@ -1216,4 +1219,25 @@ private void applyLockTime(long lockTime, Transaction tx) {
+        }
+        if (toKeep.size() != txos.size()) {
+            tx.clearOutputs();
+            for (TransactionOutput zz : toKeep) {
```
Please use a more descriptive variable name

sqrrm reviewed (changes requested):
Looks functionally correct and easy to follow. Would be helpful with a comment on how this works and what the effects are, perhaps around the `removeDust` method

Please review inline comments

Commit:
Address issues raised in review by sqrrm on 2020-03-26

- added a comment describing the `removeDust` method and its effects.
- applied a fix to the declaration of an ArrayList.
- use more descriptive variable names.
- made the logging more verbose to help log readers.

Commit:
Apply the dust fix to trades which use BSQ for the fee

Also, use `Restrictions.getMinNonDustOutput()` for the dust limit

sqrrm commented on core/src/main/java/bisq/core/btc/wallet/TradeWalletService.java:
```diff
@@ -257,6 +260,7 @@ public Transaction completeBsqTradingFeeTx(Transaction preparedBsqTx,
         checkNotNull(wallet, "Wallet must not be null");
         wallet.completeTx(sendRequest);
         Transaction resultTx = sendRequest.tx;
+        removeDust(resultTx);
```
I think this will cause BSQ change that lower than dust to be used as extra fee. That could be a non trivial amount (5.46 BSQ > USD3). While that's not a huge amount it's still too big to be an acceptable loss for users. If BSQ prices rise it will become even less so.

I think we'll have to fail transactions that try to pay the fee using BSQ if the change is less than dust. Best would be to show something at the time the fee is selected that it's likely to result in a dust change output and thus not be possible to continue. I think that should be doable, the fees are known at the time of setting up the trade so we can calculate the BSQ dust. One thing we need to do is make sure that if there is enough BSQ but split over 2 or more outputs, it's aggregated to avoid dust outputs. Example, counting satoshis:

Fee: 1000
BSQ wallet outputs:
1. 1200
2. 900

I think the tx might be constructed using only outputs 1. leaving a change output of 200, but it would be possible to construct the tx to take both 1. and 2. to pay the 1000 in fee and get the change 1100.

sqrrm replied:
It's possible the check during tx build is not there but rather later, causing it to look like a bug. Still, that's at least better than burning BSQ unwittingly. If you could check that it's handled properly and not as a bug report that would be best of course.

sqrrm reviewed (approved):
ACK
